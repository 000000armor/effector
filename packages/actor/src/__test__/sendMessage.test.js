//@flow

import {createActor, readAllIncoming} from '..'

test('send message to itself', async() => {
 const actor = createActor([{value: 'A'}])
 expect({
  state: actor.state[0],
  meta: actor.meta,
 }).toMatchSnapshot('1) initial')
 actor.after(
  (_, e) => {
   if (e === 'hello') return actor
  },
  async(actor, e) => {
   await actor.sendMessage(actor.state[0].value)
   await actor.sendMessage('how r u')
  },
 )
 await actor.sendMessage('hello')
 expect({
  state: actor.state[0],
  meta: actor.meta,
 }).toMatchSnapshot('2) hello')

 await actor.readIncoming()
 expect({
  state: actor.state[0],
  meta: actor.meta,
 }).toMatchSnapshot('3) read incoming')
})

test('to another actor', async() => {
 const actor = createActor([{value: 'A'}])
 const actorB = createActor([{value: 'B'}])
 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('1) initial')

 await actor.sendMessage('hello', actorB)
 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('2) hello')

 await actor.sendMessage('how r u', actorB)
 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('3) how r u')

 await actor.readIncoming()
 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('4) read incoming A')

 await actorB.readIncoming()
 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('5) read incoming B')
})

test('talk', async() => {
 const actor = createActor([
  {
   count: 0,
  },
  [],
 ])
 const actorB = createActor([{value: 'none'}])

 actorB.after(
  (_, e) => {
   if (e === 'hello') return actor
  },
  async(sender, e, that) => {
   await that.sendMessage('hi', sender)
   const [obj, history] = sender.state
   history.push(e)
   obj.count += 1
  },
 )
 actor.after(
  (_, e) => {
   if (e === 'hi') return actorB
  },
  async(sender, e, that) => {
   await that.sendMessage('how r u', sender)
   const [obj, history] = that.state
   history.push(e)
   obj.count += 1
  },
 )
 actorB.after(
  (_, e) => {
   if (e === 'how r u') return actor
  },
  async(sender, e, that) => {
   await that.sendMessage("I'm fine", sender)
   const [obj, history] = sender.state
   history.push(e)
   obj.count += 1
   that.state[0].value = 'fine'
  },
 )
 actor.after(
  (_, e) => {
   if (e === "I'm fine") return actorB
  },
  async(sender, e, that) => {
   await that.sendMessage('bye', sender)
   const [obj, history] = that.state
   history.push(e)
   obj.count += 1
   sender.state[0].value = 'bye'
  },
 )
 //  actor.saveHandlers()
 //  actorB.saveHandlers()
 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('1) initial')

 await actor.readIncoming()
 await actorB.readIncoming()

 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('2) read incoming (both)')

 await actor.sendMessage('hello', actorB)
 //  await actor.readIncoming()
 await actor.readIncoming()
 await actorB.readIncoming()
 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('3) hello')

 await actor.readIncoming()
 await actorB.readIncoming()
 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('4) read incoming')

 await actor.readIncoming()
 await actorB.readIncoming()
 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('5) read incoming')
})

test('readAllIncoming', async() => {
 const actor = createActor([
  {
   count: 0,
  },
  [],
 ])
 const actorB = createActor([{value: 'none'}])

 actorB.after(
  (_, e) => {
   if (e === 'hello') return actor
  },
  async(sender, e, that) => {
   await that.sendMessage('hi', sender)
   const [obj, history] = sender.state
   history.push(e)
   obj.count += 1
  },
 )
 actor.after(
  (_, e) => {
   if (e === 'hi') return actorB
  },
  async(sender, e, that) => {
   await that.sendMessage('how r u', sender)
   const [obj, history] = that.state
   history.push(e)
   obj.count += 1
  },
 )
 actorB.after(
  (_, e) => {
   if (e === 'how r u') return actor
  },
  async(sender, e, that) => {
   await that.sendMessage("I'm fine", sender)
   const [obj, history] = sender.state
   history.push(e)
   obj.count += 1
   that.state[0].value = 'fine'
  },
 )
 actor.after(
  (_, e) => {
   if (e === "I'm fine") return actorB
  },
  async(sender, e, that) => {
   await that.sendMessage('bye', sender)
   const [obj, history] = that.state
   history.push(e)
   obj.count += 1
   sender.state[0].value = 'bye'
  },
 )
 //  actor.saveHandlers()
 //  actorB.saveHandlers()
 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('1) initial')
 await readAllIncoming([actor, actorB])

 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('2) read all incoming')

 await actor.sendMessage('hello', actorB)

 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('3) hello')
 await readAllIncoming([actor, actorB])

 expect({
  state: {
   a: actor.state[0],
   b: actorB.state[0],
  },
  meta: {
   a: actor.meta,
   b: actorB.meta,
  },
 }).toMatchSnapshot('4) read all incoming')

 expect(actorB.state[0].value).toBe('bye')
})
