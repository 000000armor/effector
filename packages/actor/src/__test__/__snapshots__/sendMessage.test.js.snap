// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readAllIncoming: 1) initial 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {
        [Function],
        [Function],
      },
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {
        [Function],
        [Function],
      },
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "count": 0,
    },
    "b": Object {
      "value": "none",
    },
  },
}
`;

exports[`readAllIncoming: 2) read all incoming 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "count": 0,
    },
    "b": Object {
      "value": "none",
    },
  },
}
`;

exports[`readAllIncoming: 3) hello 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {
        "hello",
      },
    },
  },
  "state": Object {
    "a": Object {
      "count": 0,
    },
    "b": Object {
      "value": "none",
    },
  },
}
`;

exports[`readAllIncoming: 4) read all incoming 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "count": 4,
    },
    "b": Object {
      "value": "bye",
    },
  },
}
`;

exports[`send message to itself: 1) initial 1`] = `
Object {
  "meta": ActorPrivateMeta {
    "handlers": Set {},
    "incoming": Set {},
    "keepBy": null,
    "keeperFor": WeakMap {},
    "newHandlers": Set {},
    "newIncoming": Set {},
  },
  "state": Object {
    "value": "A",
  },
}
`;

exports[`send message to itself: 2) hello 1`] = `
Object {
  "meta": ActorPrivateMeta {
    "handlers": Set {},
    "incoming": Set {},
    "keepBy": null,
    "keeperFor": WeakMap {},
    "newHandlers": Set {
      [Function],
    },
    "newIncoming": Set {
      "hello",
    },
  },
  "state": Object {
    "value": "A",
  },
}
`;

exports[`send message to itself: 3) read incoming 1`] = `
Object {
  "meta": ActorPrivateMeta {
    "handlers": Set {
      [Function],
    },
    "incoming": Set {
      "A",
      "how r u",
    },
    "keepBy": null,
    "keeperFor": WeakMap {},
    "newHandlers": Set {},
    "newIncoming": Set {},
  },
  "state": Object {
    "value": "A",
  },
}
`;

exports[`talk: 1) initial 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {
        [Function],
        [Function],
      },
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {
        [Function],
        [Function],
      },
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "count": 0,
    },
    "b": Object {
      "value": "none",
    },
  },
}
`;

exports[`talk: 2) read incoming (both) 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "count": 0,
    },
    "b": Object {
      "value": "none",
    },
  },
}
`;

exports[`talk: 3) hello 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {
        "hi",
      },
    },
    "b": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "count": 1,
    },
    "b": Object {
      "value": "none",
    },
  },
}
`;

exports[`talk: 4) read incoming 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {
        "I'm fine",
      },
    },
    "b": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "count": 3,
    },
    "b": Object {
      "value": "fine",
    },
  },
}
`;

exports[`talk: 5) read incoming 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {
        [Function],
        [Function],
      },
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "count": 4,
    },
    "b": Object {
      "value": "bye",
    },
  },
}
`;

exports[`to another actor: 1) initial 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "value": "A",
    },
    "b": Object {
      "value": "B",
    },
  },
}
`;

exports[`to another actor: 2) hello 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {
        "hello",
      },
    },
  },
  "state": Object {
    "a": Object {
      "value": "A",
    },
    "b": Object {
      "value": "B",
    },
  },
}
`;

exports[`to another actor: 3) how r u 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {
        "hello",
        "how r u",
      },
    },
  },
  "state": Object {
    "a": Object {
      "value": "A",
    },
    "b": Object {
      "value": "B",
    },
  },
}
`;

exports[`to another actor: 4) read incoming A 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {
        "hello",
        "how r u",
      },
    },
  },
  "state": Object {
    "a": Object {
      "value": "A",
    },
    "b": Object {
      "value": "B",
    },
  },
}
`;

exports[`to another actor: 5) read incoming B 1`] = `
Object {
  "meta": Object {
    "a": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
    "b": ActorPrivateMeta {
      "handlers": Set {},
      "incoming": Set {},
      "keepBy": null,
      "keeperFor": WeakMap {},
      "newHandlers": Set {},
      "newIncoming": Set {},
    },
  },
  "state": Object {
    "a": Object {
      "value": "A",
    },
    "b": Object {
      "value": "B",
    },
  },
}
`;
